hardfault.s

```diff
@@ -0,0 +1,16 @@
+  .global HardFault
+  .thumb_func
+HardFault:
+  movs r0, #4
+  mov  r1, lr
+  tst  r0, r1      // Test bit[3] of EXC_RETURN to determine thread mode
```

Comment:
Shouldn't this say "bit[2]"? Bit positions start at 0 to have bitmasks match `1 << $pos` (bit 0 -> mask = (1 << 0); bit 1 -> mask = (1 << 1); etc.)

---

I think you can use `bmi` instead of `tst` to use less registers. For example, this code

``` rust
#![no_std]
#![feature(asm)]
#![feature(core_intrinsics)]

use core::intrinsics;

pub fn foo(x: u32) {
    unsafe {
        if x & (1 << 2) == 0 {
            asm!("mrs r0, MSP
                  bl UserHardFault" :::: "volatile");
            intrinsics::unreachable();
        } else {
            asm!("mrs r0, PSP
                  bl UserHardFault" :::: "volatile");
            intrinsics::unreachable();
        }
    }
}
```

produces this machine code when compiled with optimizations:

``` armasm
00000000 <foo::foo>:
   0:   0740            lsls    r0, r0, #29
   2:   d404            bmi.n   e <foo::foo+0xe>
   4:   f3ef 8008       mrs     r0, MSP
   8:   f7ff fffe       bl      0 <UserHardFault>
   c:   defe            udf     #254    ; 0xfe
   e:   f3ef 8009       mrs     r0, PSP
  12:   f7ff fffe       bl      0 <UserHardFault>
  16:   defe            udf     #254    ; 0xfe
```

The `defe udf` instruction can be omitted.